voice/src/main/kotlin/Streams.kt

```diff
@@ -0,0 +1,76 @@
+package dev.kord.voice
+
+import dev.kord.common.annotation.KordVoice
+import dev.kord.common.entity.Snowflake
+import dev.kord.voice.gateway.Speaking
+import dev.kord.voice.gateway.on
+import dev.kord.voice.udp.AudioPacket
+import kotlinx.atomicfu.AtomicRef
+import kotlinx.atomicfu.atomic
+import kotlinx.atomicfu.update
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.flow.*
+import kotlin.coroutines.CoroutineContext
+
+@KordVoice
+class Streams(
+    connection: VoiceConnection,
+    dispatcher: CoroutineDispatcher
+) : CoroutineScope {
+    override val coroutineContext: CoroutineContext =
+        SupervisorJob() + dispatcher + CoroutineName("Voice Connection Incoming Streams")
+
+    internal val key: AtomicRef<ByteArray?> = atomic(null)
+
+    /**
+     * A flow of all incoming [dev.kord.voice.udp.AudioPacket.DecryptedPacket]s through the UDP connection.
+     */
+    val incomingAudioPackets: SharedFlow<AudioPacket.DecryptedPacket> =
+        connection.udp
+            .incoming
+            .mapNotNull(AudioPacket::encryptedFrom)
+            .map { it.decrypt(key.value!!) }
+            .shareIn(this, SharingStarted.Lazily)
+
+    /**
+     * A flow of all incoming [AudioFrame]s mapped to their ssrc.
+     */
+    val incomingAudioFrames get() = incomingAudioPackets.map { it.ssrc to AudioFrame.fromData(it.data)!! }
+
+    private val _incomingUserAudioFrames: MutableSharedFlow<Pair<Snowflake, AudioFrame>> = MutableSharedFlow()
+
+    /**
+     * A flow of incoming [AudioFrame]s mapped to its corresponding user id. Streams for each user are built over time,
+     * or whenever the [dev.kord.voice.gateway.VoiceGateway] receives a [Speaking] event.
+     */
+    val incomingUserStreams: SharedFlow<Pair<Snowflake, AudioFrame>> = _incomingUserAudioFrames
```

Comment:
`incomingUserStreams` could also use `_incomingUserAudioFrames.asSharedFlow()` to forbid casting

Reply:
This pattern is used a few other places throughout Kord, is there any other benefit to do this?

Reply:
Nothing I know about, this is not a must (but I thought I could mention it).